### PR TITLE
servicebus jms connection-string enhance

### PR DESCRIPTION
**Pull request by jialigit**
# Description

 For now, serivcebus jms starter's credential can only be configured using the configuration option of "spring.jms.servicebus.connection-string". We also need to support the option of :
1.   configuration option of "spring.cloud.azure.servicebus.connection-string". 
2.  configuration of  "spring.cloud.azure.servicebus.namespace" used with "credential" option, i.e.  
the following configuration can be used.
```
#
spring:
  jms:
    servicebus:
      connection-string:  ${AZURE_SERVICEBUS_CONNECTION_STRING}
```
and 
```application-sp.yml
spring:
  cloud:
    azure:
      servicebus:
        connection-string:  ${AZURE_SERVICEBUS_CONNECTION_STRING}
 ...
```

OR

```
spring:
  cloud:
    azure:
      credential:
        client-id:  ${AZURE_CLIENT_ID}
        client-secret:  ${AZURE_CLIENT_SECRET}
      profile:
          tenant-id:  ${AZURE_TENANT_ID}
          subscription-id:  ${SUBSCRIPTION_ID}
      servicebus:
        resource:
          resource-group:  ${RESOURCE_GROUP}
        namespace: ${NAMESPACE}


```

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](../../CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java [hunk 1]
```diff
@@ -30,8 +38,11 @@
  */
 @Configuration(proxyBeanMethods = false)
 @AutoConfigureBefore(JmsAutoConfiguration.class)
-@AutoConfigureAfter(JndiConnectionFactoryAutoConfiguration.class)
+@AutoConfigureAfter({
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java:41]: will the annotation @AutoConfigureBefore @AutoConfigureAfter work here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java [hunk 2]
```diff
@@ -47,4 +58,29 @@ ServiceBusJmsConnectionFactoryCustomizer amqpOpenPropertiesCustomizer() {
...
+            HasConnectionStringProviderConfigured() {
+            }
+        }
+
+        @ConditionalOnProperty(
+            prefix = "spring.jms.servicebus",
+            name = { "connection-string" }
+        )
+        static class HasSpringJmsServiceBusConfigured {
+            HasSpringJmsServiceBusConfigured() {
```
Comment (backwind1233) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java:82]: do we need the constructor here?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/AzureServiceBusJmsPropertiesBeanPostProcessor.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    AzureServiceBusJmsPropertiesBeanPostProcessor(ObjectProvider<ConnectionStringProvider<AzureServiceType.ServiceBus>> connectionStringProviders) {
+        this.connectionStringProviders = connectionStringProviders;
+    }
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+        if (bean instanceof AzureServiceBusJmsProperties) {
+            AzureServiceBusJmsProperties jmsProperties = (AzureServiceBusJmsProperties) bean;
+            connectionStringProviders.ifAvailable(provider -> jmsProperties.setConnectionString(provider.getConnectionString()));
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/AzureServiceBusJmsPropertiesBeanPostProcessor.java:30]: This should be called only when the connections string is null in AzureServiceBusJmsProperties.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/AzureServiceBusJmsPropertiesBeanPostProcessor.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                new ServiceBusConnectionString(connectionString);
+            String host = serviceBusConnectionString.getEndpointUri().getHost();
+            String remoteUrl = String.format(AzureServiceBusJmsProperties.AMQP_URI_FORMAT, host,
+                jmsProperties.getIdleTimeout().toMillis());
+            String username = serviceBusConnectionString.getSharedAccessKeyName();
+            String password = serviceBusConnectionString.getSharedAccessKey();
+
+            jmsProperties.setRemoteUrl(remoteUrl);
+            jmsProperties.setUsername(username);
+            jmsProperties.setPassword(password);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/AzureServiceBusJmsPropertiesBeanPostProcessor.java:43]: Do we still need this? I think the afterPropertiesSet will handle this?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java [hunk 5]
```diff
@@ -47,4 +60,31 @@ ServiceBusJmsConnectionFactoryCustomizer amqpOpenPropertiesCustomizer() {
                 (connection, uri) -> properties);
         };
     }
+
+    @Bean
+    @ConditionalOnMissingBean
+    AzureServiceBusJmsPropertiesBeanPostProcessor azureServiceBusJmsPropertiesBeanPostProcessor(ObjectProvider<ConnectionStringProvider<AzureServiceType.ServiceBus>> connectionStringProviders) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java:66]: Does this need to be static?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java [hunk 6]
```diff
@@ -30,8 +40,11 @@
  */
 @Configuration(proxyBeanMethods = false)
 @AutoConfigureBefore(JmsAutoConfiguration.class)
-@AutoConfigureAfter(JndiConnectionFactoryAutoConfiguration.class)
+@AutoConfigureAfter({
+    JndiConnectionFactoryAutoConfiguration.class,
+    AzureServiceBusResourceManagerAutoConfiguration.class })
 @ConditionalOnClass({ ConnectionFactory.class, JmsConnectionFactory.class, JmsTemplate.class })
+@Conditional(ServiceBusJmsAutoConfiguration.OnConnectionStringProvidedCondition.class)
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java:47]: I think we don't need this anymore, the afterpropertiesset will throw exception

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 7]
```diff
@@ -26,23 +26,29 @@
...
-                                                                             ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
-        return new ServiceBusJmsConnectionFactoryFactory(properties,
-            factoryCustomizers.orderedStream().collect(Collectors.toList()))
+    @Bean
+    @ConditionalOnMissingBean
+    ServiceBusJmsConnectionFactoryFactory serviceBusJmsConnectionFactoryFactory(AzureServiceBusJmsProperties jmsProperties,
+                                                                                ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
+        return new ServiceBusJmsConnectionFactoryFactory(jmsProperties,
+            factoryCustomizers.orderedStream().collect(Collectors.toList()));
+    }
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java:35]: No need to be a bean anymore

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 8]
```diff
@@ -26,23 +26,29 @@
...
+
+    private static ServiceBusJmsConnectionFactory createJmsConnectionFactory(ServiceBusJmsConnectionFactoryFactory serviceBusJmsConnectionFactoryFactory) {
+        return serviceBusJmsConnectionFactoryFactory
             .createConnectionFactory(ServiceBusJmsConnectionFactory.class);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java:39]: No need for this method

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 9]
```diff
@@ -26,23 +26,29 @@
...
         matchIfMissing = true)
     static class SimpleConnectionFactoryConfiguration {
 
         @Bean
         @ConditionalOnProperty(prefix = "spring.jms.cache", name = "enabled", havingValue = "false")
-        ServiceBusJmsConnectionFactory jmsConnectionFactory(AzureServiceBusJmsProperties properties,
-                                                            ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
-            return createJmsConnectionFactory(properties, factoryCustomizers);
+        ServiceBusJmsConnectionFactory jmsConnectionFactory(ServiceBusJmsConnectionFactoryFactory serviceBusJmsConnectionFactoryFactory) {
+            return createJmsConnectionFactory(serviceBusJmsConnectionFactoryFactory);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java:51]: Let's keep this unchanged.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 10]
```diff
@@ -53,9 +59,9 @@ ServiceBusJmsConnectionFactory jmsConnectionFactory(AzureServiceBusJmsProperties
 
             @Bean
             CachingConnectionFactory jmsConnectionFactory(JmsProperties jmsProperties,
-                                                          AzureServiceBusJmsProperties properties,
-                                                          ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
-                ServiceBusJmsConnectionFactory factory = createJmsConnectionFactory(properties, factoryCustomizers);
+                                                          ServiceBusJmsConnectionFactoryFactory serviceBusJmsConnectionFactoryFactory) {
+                ServiceBusJmsConnectionFactory factory =
+                    createJmsConnectionFactory(serviceBusJmsConnectionFactoryFactory);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java:64]: same here

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java [hunk 11]
```diff
@@ -64,18 +70,18 @@ CachingConnectionFactory jmsConnectionFactory(JmsProperties jmsProperties,
...
     @ConditionalOnClass({ JmsPoolConnectionFactory.class, PooledObject.class })
     static class PooledConnectionFactoryConfiguration {
 
         @Bean(destroyMethod = "stop")
         @ConditionalOnProperty(prefix = "spring.jms.servicebus.pool", name = "enabled", havingValue = "true")
         JmsPoolConnectionFactory jmsPoolConnectionFactory(AzureServiceBusJmsProperties properties,
-                                                          ObjectProvider<ServiceBusJmsConnectionFactoryCustomizer> factoryCustomizers) {
-            ServiceBusJmsConnectionFactory factory = createJmsConnectionFactory(properties, factoryCustomizers);
+                                                          ServiceBusJmsConnectionFactoryFactory serviceBusJmsConnectionFactoryFactory) {
+            ServiceBusJmsConnectionFactory factory = createJmsConnectionFactory(serviceBusJmsConnectionFactoryFactory);
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsConnectionFactoryConfiguration.java:84]: same here

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/properties/AzureServiceBusJmsProperties.java [hunk 12]
```diff
@@ -23,7 +21,7 @@
 
     private static final String DEFAULT_REMOTE_URL = "amqp://localhost:5672";
 
-    private static final String AMQP_URI_FORMAT = "amqps://%s?amqp.idleTimeout=%d";
+    public static final String AMQP_URI_FORMAT = "amqps://%s?amqp.idleTimeout=%d";
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/properties/AzureServiceBusJmsProperties.java:24]: No need to change this

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/properties/AzureServiceBusJmsProperties.java [hunk 13]
```diff
@@ -185,20 +183,9 @@ public PrefetchPolicy getPrefetchPolicy() {
...
         if (null == pricingTier || !pricingTier.matches("(?i)premium|standard|basic")) {
             throw new IllegalArgumentException("'spring.jms.servicebus.pricing-tier' is not valid");
         }
-
-        ServiceBusConnectionString serviceBusConnectionString = new ServiceBusConnectionString(connectionString);
-        String host = serviceBusConnectionString.getEndpointUri().getHost();
-
-        this.remoteUrl = String.format(AMQP_URI_FORMAT, host, idleTimeout.toMillis());
-        this.username = serviceBusConnectionString.getSharedAccessKeyName();
-        this.password = serviceBusConnectionString.getSharedAccessKey();
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/properties/AzureServiceBusJmsProperties.java:201]: We can still keep them here

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java (hunk 2 above)
Comment (jialigit) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java]: refactored

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java [hunk 14]
```diff
@@ -47,4 +55,11 @@ ServiceBusJmsConnectionFactoryCustomizer amqpOpenPropertiesCustomizer() {
                 (connection, uri) -> properties);
         };
     }
+
+    @Bean
+    @ConditionalOnMissingBean
+    @ConditionalOnMissingProperty(prefix = "spring.jms.servicebus", name = "connection-string")
+    AzureServiceBusJmsPropertiesBeanPostProcessor azureServiceBusJmsPropertiesBeanPostProcessor(ObjectProvider<ConnectionStringProvider<AzureServiceType.ServiceBus>> connectionStringProviders) {
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java:62]: Should this bean be static?

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfigurationTests.java [hunk 15]
```diff
@@ -149,6 +139,18 @@ void autoconfigurationEnabledAndContextSuccessWithPremiumTier(String pricingTier
...
+    void doesnotHaveBeanOfAzureServiceBusJmsPropertiesBeanPostProcessor(String pricingTier) {
+        this.contextRunner
+            .withPropertyValues(
+                "spring.jms.servicebus.pricing-tier=" + pricingTier,
+                "spring.jms.servicebus.connection-string=" + CONNECTION_STRING)
+            .run(context -> {
+                assertThat(context).doesNotHaveBean(AzureServiceBusJmsPropertiesBeanPostProcessor.class);
+            });
+    }
+
```
Comment (saragluna) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfigurationTests.java:153]: What about the case of having a post processor?

Comment (saragluna) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfigurationTests.java:153]: And should we have a test case to make sure the post processor is working?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfigurationTests.java (hunk 15 above)
Comment (jialigit) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfigurationTests.java:158]: yes, it is a good idea.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java (hunk 14 above)
Comment (jialigit) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/jms/ServiceBusJmsAutoConfiguration.java]: yes, it makes sense.

**Review by saragluna — APPROVED** [review]
LGTM.

**Pull request merged**